pkg/cloudprovider/provider/alibaba/provider.go

```diff
@@ -0,0 +1,413 @@
+	}
+	if c.InternetMaxBandwidthOut == "" {
+		return fmt.Errorf("internetMaxBandwidthOut is missing")
+	}
+	if c.ZoneID == "" {
+		return fmt.Errorf("zoneID is missing")
+	}
+	_, err = getImageIDForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	return nil
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData) (instance.Instance, error) {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client, err := getClient(c.RegionID, c.AccessKeyID, c.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+
+	i, err := getInstance(client, machine.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if i != nil {
+		return &alibabaInstance{instance: i}, nil
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
+
+func (p *provider) GetCloudConfig(spec v1alpha1.MachineSpec) (config string, name string, err error) {
+	return "", "", nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client, err := getClient(c.RegionID, c.AccessKeyID, c.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.Labels == nil {
+		c.Labels = map[string]string{}
+	}
+
+	c.Labels[machineUIDTag] = string(machine.UID)
+
+	createInstanceRequest := ecs.CreateCreateInstanceRequest()
+	createInstanceRequest.ImageId, _ = getImageIDForOS(pc.OperatingSystem)
+	createInstanceRequest.InstanceName = machine.Name
+	createInstanceRequest.InstanceType = c.InstanceType
+	createInstanceRequest.VSwitchId = c.VSwitchID
+	createInstanceRequest.InternetMaxBandwidthOut = requests.Integer(c.InternetMaxBandwidthOut)
+	encodedUserData := base64.StdEncoding.EncodeToString([]byte(userdata))
+	createInstanceRequest.UserData = encodedUserData
+	createInstanceRequest.SystemDiskCategory = "cloud_efficiency"
+	createInstanceRequest.ZoneId = c.ZoneID
+
+	res, err := client.CreateInstance(createInstanceRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create instance at Alibaba cloud: %v", err)
+	}
+
+	var (
+		foundInstance *ecs.Instance
+		maxRetries    = 10
+		tries         = 0
+	)
+
+	for {
+		foundInstance, err = getInstance(client, machine.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		if tries >= maxRetries {
+			return nil, fmt.Errorf("instance %v isn't in a ready state", foundInstance.InstanceId)
+		}
+
+		switch instanceStatus(foundInstance.Status) {
```

Comment:
Blocking for a prolongued time is a bit problematic, can it be avoided, e.G. by starting the instance within the `Get` func?